src/gui/inputdevices/TouchInputHandler.cpp

```diff
@@ -137,16 +132,25 @@ void TouchInputHandler::zoomStart() {
         zoomControl->setZoomFitMode(false);
     }
 
+    auto center = (this->priLastAbs + this->secLastAbs) / 2.0;
+
     auto* mainWindow = inputContext->getView()->getControl()->getWindow();
+    // use screen pixel coordinates for the zoom center
+    // as relative coordinates depend on the changing zoom level
+    int rx, ry;
+    gdk_window_get_root_coords(*mainWindow, 0, 0, &rx, &ry);
```

Comment:
This is wrong. The mainWindow here is of type MainWindow, not GtkWindow. We need the gtk window of the scroll widget, which seems to be a bit harder to get. Otherwise this will return {0,0} instead of the right root_coords which causes a little offset in the zoom center. 

This fixes that but without caring about the touch workaround (in that case, another GtkWidget seems to be used in MainWindow). Also, maybe there is another way of getting the right widget/root coords without adding an additional method to MainWindow.
```diff
diff --git a/src/gui/MainWindow.cpp b/src/gui/MainWindow.cpp
index 3e03ee08..6dc30672 100644
--- a/src/gui/MainWindow.cpp
+++ b/src/gui/MainWindow.cpp
@@ -340,6 +340,8 @@ void MainWindow::initHideMenu() {
 
 auto MainWindow::getLayout() -> Layout* { return gtk_xournal_get_layout(GTK_WIDGET(this->xournal->getWidget())); }
 
+auto MainWindow::getWinXournal() -> GtkWidget* { return this->winXournal; }
+
 auto cancellable_cancel(GCancellable* cancel) -> bool {
     g_cancellable_cancel(cancel);
 
diff --git a/src/gui/MainWindow.h b/src/gui/MainWindow.h
index 875ac3fc..bc8245fc 100644
--- a/src/gui/MainWindow.h
+++ b/src/gui/MainWindow.h
@@ -98,6 +98,7 @@ public:
 
     bool isGestureActive();
 
+    GtkWidget* getWinXournal();
 
     /**
      * Disable kinetic scrolling if there is a touchscreen device that was manually mapped to another enabled input
diff --git a/src/gui/inputdevices/TouchInputHandler.cpp b/src/gui/inputdevices/TouchInputHandler.cpp
index 73a0737a..9573ebbe 100644
--- a/src/gui/inputdevices/TouchInputHandler.cpp
+++ b/src/gui/inputdevices/TouchInputHandler.cpp
@@ -138,10 +138,11 @@ void TouchInputHandler::zoomStart() {
     // use screen pixel coordinates for the zoom center
     // as relative coordinates depend on the changing zoom level
     int rx, ry;
-    gdk_window_get_root_coords(*mainWindow, 0, 0, &rx, &ry);
+    gdk_window_get_root_coords(gtk_widget_get_window(mainWindow->getWinXournal()), 0, 0, &rx, &ry);
     center.x -= rx;
     center.y -= ry;

     // When not using touch drawing, we're using a different scrolling method.
     // This requires different centering.
     /*if (!mainWindow->getGtkTouchscreenScrollingEnabled()) {
 ```